torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -173,40 +173,50 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   }
 
   /// Adds a new (boxed) `Module` to the `Sequential` container.
+  ///
+  /// See `add_to_modules(std::shared_ptr<ModuleType> module_ptr)` for
+  /// implementation detail.
   template <typename ModuleType>
   void push_back(std::shared_ptr<ModuleType> module_ptr) {
-    // Nesting Sequential doesn't work because `forward()`'s return type is
-    // templatized, so it'll give a nasty compiler error.
-    static_assert(
-        !std::is_same<SequentialImpl, ModuleType>::value,
-        "Sequential is not nestable");
-    static_assert(
-        torch::detail::is_module<ModuleType>::value,
-        "Can only add objects derived from nn::Module to Sequential");
-    static_assert(
-        torch::detail::has_forward<ModuleType>::value,
-        "Can only add modules with a forward() method to Sequential");
-    push_back(AnyModule(std::move(module_ptr)));
+    auto index = add_to_modules(module_ptr);
+    register_module(std::to_string(index), modules_[index].ptr());
   }
 
-  /// Adds a new `Module` to the `Sequential` container, moving or copying it
-  /// into a `shared_ptr` internally. This method allows passing value types,
-  /// and letting the container deal with the boxing. This means you can write
+  /// Adds a new `Module` to the `Sequential` container.
+  /// This method allows passing value types, and letting the container
+  /// deal with the boxing. This means you can write
   /// `Sequential(Module(3, 4))` instead of
   /// `Sequential(std::make_shared<Module>(3, 4))`.
+  ///
+  /// See `add_to_modules(M&& module)` for implementation detail.
   template <typename M, typename = torch::detail::enable_if_module_t<M>>
   void push_back(M&& module) {
-    // Need to get rid of any reference components for make_unique.
-    using Type = typename std::remove_reference<M>::type;
-    // Here we move (or copy) the module into a new shared_ptr.
-    push_back(std::make_shared<Type>(std::forward<M>(module)));
+    auto index = add_to_modules(module);
+    register_module(std::to_string(index), modules_[index].ptr());
   }
 
   /// Unwraps the contained module of a `ModuleHolder` and adds it to the
   /// `Sequential`.
+  ///
+  /// See `add_to_modules(const ModuleHolder<M>& module_holder)` for implementation detail.
   template <typename M>
   void push_back(const ModuleHolder<M>& module_holder) {
-    push_back(module_holder.ptr());
+    auto index = add_to_modules(module_holder);
+    register_module(std::to_string(index), modules_[index].ptr());
+  }
+
+  /// Adds a new named module to the `Sequential` container, with name of `std::string` type.
+  template <typename M>
+  void push_back(std::string name, M module) {
+    auto index = add_to_modules(module);
+    register_module(name, modules_[index].ptr());
+  }
+
+  /// Adds a new named module to the `Sequential` container, with name of `const char*` type.
+  template <typename M>
+  void push_back(const char* name, M module) {
```

Comment:
I believe this does an extra copy of the module unnecessarily, and doesn't really follow the convention set above. I'm not sure how expensive the copy constructor is for Module.

It may be better to follow the construct given above, having a separate function for each of the possible types, i.e.
- std::shared_ptr<ModuleType>
- const ModuleHolder<M>&
- M&& with M being a module

Alternatively if you think the construct doesn't matter and the extra copy is whatever, then we can simplify the above code, but I think it's important to be consistent.